### PR TITLE
fix: ramen test should wait worker stats

### LINKED ISSUE #1254: ramen test should wait worker stats

### PR DESCRIPTION
**Pull request by darlentar**
fix #1254

### CODE REVIEW

File: src/RamenTests.ml [hunk 1]
```diff
@@ -437,7 +437,26 @@ let process_until what ~while_ session cond =
 (* Read all workers stats directly from the confserver internal hash (ie.
  * confserver must run in this program).
  * Returns a hash of N.fq to runtime stats *)
-let read_stats session =
+let read_stats session now while_ =
+  let null_stats = ref true in
+  let stats_thread = Thread.create (fun () ->
+    let null_stats_i = ref true in
+    while !null_stats_i; do
```
Comment (rixed) [inline src/RamenTests.ml:444]: I think I get the idea, but there are several problems with this implementation:

- First and foremost, indentation is bogus :)
- Second, the style is needlessly imperative
- Then, this is busy looping!
- Finally, I'm not convinced that a separate thread is needed here. Why not wait for the stat reports in the main current thread?

File: src/RamenTests.ml [hunk 2]
```diff
@@ -583,6 +602,8 @@ let run conf server_url api graphite
    * Start all services as threads
    *)
   let while_ () = !RamenProcesses.quit = None in
+  let test_is_not_finished = ref true in
```
Comment (rixed) [inline src/RamenTests.ml:605]: Let's avoid double negations

File: src/RamenTests.ml [hunk 3]
```diff
@@ -650,27 +674,28 @@ let run conf server_url api graphite
...
   if ok then !logger.info "Test %s: Success" name
   else !logger.error "Test %s: FAILURE" name ;
   if httpd_thread = None then
     RamenProcesses.quit := Some 0 ;
-  let join_thread what thd =
-    !logger.debug "Waiting for %s termination..." what ;
-    Thread.join thd in
   (* else wait for the user to kill *)
   Option.may (join_thread "httpd") httpd_thread ;
-  join_thread "supervisor" supervisor_thread ;
```
Comment (rixed) [inline src/RamenTests.ml:673]: Why not?

**Review by rixed — CHANGES_REQUESTED** [review]

File: src/RamenTests.ml (hunk 3 above)
Comment (darlentar) [reply src/RamenTests.ml:673]: ? I wait the thread when I want to terminate the supervisor just a few lines above.

File: src/RamenTests.ml (hunk 1 above)
Comment (darlentar) [reply src/RamenTests.ml:444]: First -> Ok.
Second -> Yes, it is really a problem ? Would you prefer a rec loop ?
Third -> Right. I could add a sleepf(500) WDYT? 
Fourth -> ? I need to read the values with `process_until` in the same time so I need a thread.

Comment (rixed) [reply src/RamenTests.ml]: Process_until is just a loop calling process_in until some condition is met.
Your condition is that we have fresh stats for all workers.
Can't you just do: `process_until "stats_are_received" ~while_ session all_stats_are_fresh` with `all_stats_are_fresh : unit -> bool` returning true if every stat report's stats_time is after `now` (or some timeout) ?

File: src/RamenTests.ml (hunk 3 above)
Comment (rixed) [reply src/RamenTests.ml:673]: true

File: src/CodeGenLib_Skeletons.ml [hunk 4]
```diff
@@ -226,14 +226,17 @@ let worker_start conf get_binocle_tuple
...
   match k publish_stats outputer with
   | exception e ->
       print_exception ~what:"Worker process" e ;
       last_report () ;
+      Option.may (Thread.join) (!Publish.thd) ;
       exit ExitCodes.uncaught_exception
   | () ->
       last_report () ;
+      Option.may (Thread.join) (!Publish.thd) ;
       exit (!quit |? ExitCodes.terminated)
```
Comment (rixed) [inline src/CodeGenLib_Skeletons.ml:240]: Wouldn't that make sense to have this fix in a distinct commit?

File: src/RamenSupervisor.ml [hunk 5]
```diff
@@ -943,24 +943,43 @@ let synchronize_once =
...
+    let pid, st = Unix.wait() in
+    if pid = -1 then !logger.info "Some error as happend in waitpid" ;
+    if pid != 0 then (
+      match st with
+        | WEXITED i -> !logger.info "Exist with %d" i
+        | WSIGNALED i -> !logger.info "Killed by signal %d" i
+        | WSTOPPED i -> !logger.info "Stopped by signal %d" i) ;
+    let pids = List.remove pids pid in
+    wait_pids pids)
+  in
```
Comment (rixed) [inline src/RamenSupervisor.ml:970]: What's wrong with `RamenHelpers.waitall_once` ?

File: src/RamenSupervisor.ml [hunk 6]
```diff
@@ -943,24 +943,43 @@ let synchronize_once =
...
+
 
 (* At start, assume pre-existing pids are left-overs from a previous run, but
  * if synchronize_running is restarted because of some exception then do not
  * assume any longer that previous workers are not running: *)
 let previous_pids_are_running = ref false
 
-let synchronize_running conf kill_at_exit =
-  let while_ () = !Processes.quit = None in
+let synchronize_running ?(while_=fun () -> !Processes.quit = None)
```
Comment (rixed) [inline src/RamenSupervisor.ml:981]: Let's not make a special case: use `always` as the default value for while_.

File: src/RamenTests.ml [hunk 7]
```diff
@@ -645,32 +673,36 @@ let run conf server_url api graphite
       "sites/*/workers/*/outputs" ;
       (* To display resources *)
       "sites/*/workers/*/stats/runtime" ] in
-  let ok, stats =
+  let ok, max_retries_tested, stats =
```
Comment (rixed) [inline src/RamenTests.ml:676]: Once you have that `all_stats_are_fresh` function you can do away with this inelegant `max_retries_tested`.

File: src/RamenTests.ml (hunk 1 above)
Comment (darlentar) [reply src/RamenTests.ml]: We can't currently do this with the current implementation of `process_until` as we need a way to update the stats state. 
Currently code of `process_until` (well it's `process_in` actually) is:
```
      | msg ->
          Client.process_msg session.clt msg ;
          if not single then loop ()
```
We do not do anything but reading incoming messages from the server. A solution could be to add a `process ` parameter and do something like
```diff
       | msg ->
           Client.process_msg session.clt msg ;
+          process msg ;
           if not single then loop ()
```

WDYT ?

File: src/RamenSupervisor.ml (hunk 5 above)
Comment (darlentar) [reply src/RamenSupervisor.ml:970]: I just don't see these function but indeed it works.

File: src/CodeGenLib_Skeletons.ml (hunk 4 above)
Comment (darlentar) [reply src/CodeGenLib_Skeletons.ml:240]: I don't know if you want too. I put the fix in this commit as the bug is revealed by the fact I'm waiting the stats. I can do another commit though.

File: src/RamenTests.ml (hunk 1 above)
Comment (darlentar) [reply src/RamenTests.ml]: Oh i see. I will do this !

**Comment by rixed** [comment]
Your new approach is too sophisticated for this fix.

Once the workers do publish their stats before quitting (there was yet another fix required for that), the solution is rather simple: just terminate confserver last, after all received stats have a timestamp greater than the end of the tests (not really required in practice, just to be extra secure).
That's quite similar to your original approach, with no additional thread needed.

I've done this in master and checked that it worked, therefore closing.

**Pull request closed without merging**